sumdbaudit/audit/service.go

```diff
@@ -184,10 +193,107 @@ func (s *Service) HashTiles(ctx context.Context, checkpoint *tlog.Tree) error {
+		}
+	}
+	head, err := s.localDB.Head()
+	if err != nil {
+		return fmt.Errorf("cannot find head of local log: %v", err)
+	}
+	if golden.N > head {
+		// This can happen if SumDB hasn't committed to any new complete tiles since the
+		// last run of the auditor. Any changes in tile hashes will be caught by HashTiles
+		// so skipping the consistency check is safer than it appears.
+		glog.Infof("golden STH is at size %d but local log only has %d entries; consistency check skipped", golden.N, head)
+		return nil
+	}
+
+	logRange, stragglersCount, err := s.hashAtSize(golden.N)
+	if err != nil {
+		return fmt.Errorf("failed to hash log: %v", err)
+	}
+	if stragglersCount > 0 {
+		stragglerTileOffset := int(golden.N / (1 << s.height))
+		tHashes, err := s.localDB.Tile(s.height, 0, stragglerTileOffset)
+		if err != nil {
+			return fmt.Errorf("failed to get tile: %w", err)
+		}
+		for i := 0; i < stragglersCount; i++ {
+			logRange.Append(tHashes[i], nil)
+		}
+	}
+	if logRange.End() != uint64(golden.N) {
+		return fmt.Errorf("calculation error, found %d leaves but expected %d", logRange.End(), golden.N)
+	}
+
+	root, err := logRange.GetRootHash(nil)
+	if err != nil {
+		return fmt.Errorf("failed to get root hash: %v", err)
+	}
+	var rootHash tlog.Hash
+	copy(rootHash[:], root)
+	if rootHash != golden.Hash {
+		return fmt.Errorf("log root mismatch at tree size %d; calculated %x, SumDB says %x", golden.N, root, golden.Hash[:])
+	}
+	return nil
+}
+
 // CheckRootHash calculates the root hash from the locally generated tiles, and then
 // appends any stragglers from the SumDB, returning an error if this calculation
 // fails or the result does not match that in the checkpoint provided.
-func (s *Service) CheckRootHash(ctx context.Context, checkpoint *tlog.Tree) error {
+func (s *Service) CheckRootHash(ctx context.Context, checkpoint *Checkpoint) error {
+	logRange, stragglersCount, err := s.hashAtSize(checkpoint.N)
+	if err != nil {
+		return fmt.Errorf("failed to hash log: %v", err)
+	}
+	// All full tiles have now been computed. Now we need to add the stragglers to the
+	// running hash in `logRange` in order to compare with `checkpoint`.
+	stragglerTileOffset := int(checkpoint.N / (1 << s.height))
+	stragglers, err := s.sumDB.PartialLeavesAtOffset(stragglerTileOffset, stragglersCount)
+	if err != nil {
+		return fmt.Errorf("failed to get stragglers: %v", err)
+	}
+	for _, s := range stragglers {
+		sHash := tlog.RecordHash(s)
+		logRange.Append(sHash[:], nil)
+	}
+
+	if logRange.End() != uint64(checkpoint.N) {
+		return fmt.Errorf("calculation error, found %d leaves but expected %d", logRange.End(), checkpoint.N)
+	}
+
+	root, err := logRange.GetRootHash(nil)
+	if err != nil {
+		return fmt.Errorf("failed to get root hash: %v", err)
+	}
+	var rootHash tlog.Hash
+	copy(rootHash[:], root)
+	if rootHash != checkpoint.Hash {
+		return fmt.Errorf("log root mismatch at tree size %d; calculated %x, SumDB says %x", checkpoint.N, root, checkpoint.Hash[:])
+	}
```

Comment:
There's a bit of overlap at the top & tail of this method with the `CheckConsistency` method, is it possible/would it make sense to factor out and reuse?

Reply:
Refactored the straggler processing into this method via a function to get them.